### PR TITLE
Ssteel/tensorizer support

### PR DESCRIPTION
**Pull request by sangstar**
# Tensorizer support for `vllm` library
Allows model serving using `vllm.entrypoints.openai.api_server.py` to load models using CoreWeave's [Tensorizer](https://github.com/coreweave/tensorizer). Adds `tensorizer` as a `load_format`, and deserializes a model from a S3 endpoint URI or local path from `download_dir`.

## Example:
```
python3 -m vllm.entrypoints.openai.api_server --host "0.0.0.0" \
--model "mistralai/Mistral-7B-v0.1" \
--load-format "tensorizer" \
--download-dir "s3://tensorized-ssteel/Mistral-7B-v0.1-vllm-alternative.tensors" \
--plaid-mode \
```

## Caveats:
- User must be authenticated with S3 and have read/write access to the endpoint

## Models tested on
- `meta-llama/Llama-2-7b-hf`
- `EleutherAI/gpt-j-6b`
- `mistralai/Mistral-7B-v0.1`
- `tiiuae/falcon-7b`
- `01-ai/Yi-6B`

## To do:

- [x] Allow args for `TensorDeserializer` to be passed through
- [x] Reduce code smell
- [x] Test for multiple tensorized models
- [ ] Add error-handling for edge cases (such as throwing an error if `tensorizer_path` is not provided when `load_format="tensorizer"`, if `tensorizer_path` is provided but `load_format != "tensorizer"` etc)
- [x] Resolve changes from dmarx
- [x] Test model load by deserializing from a local file path rather than S3 URI
- [x] If model weights are serialized and uploaded because `download_dir` was not found, provide functionality to allow for model serving to still occur with the downloaded HF weights
- [ ] Give better defaults for serialization and deserialization (like `plaid_mode = True` likely)
- [ ] Revisit implementation; whether in fact not appropriate to introduce new config object, if implementation of slotting `tensorizer_args` into `ModelConfig` is appropriate etc
- [x] Formatting, uncommenting, removing print debugging statements etc
- [ ] Implementing encryption

### CODE REVIEW

File: vllm/config.py [hunk 1]
```diff
@@ -55,6 +56,9 @@ class ModelConfig:
         max_context_len_to_capture: Maximum context len covered by CUDA graphs.
             When a sequence has context length larger than this, we fall back
             to eager mode.
+        tensorizer_path: File path or s3 URI to the tensorizer model weights
```
Comment (dmarx) [inline vllm/config.py]: i'm wondering if maybe it would make sense to replace this argument with a dict or dataclass (`TensorizerConfig`?) of misc TensorDeserializer args. this would facilitate leveraging additional tensorizer functionality like encryption or plaid mode.

File: vllm/model_executor/weight_utils.py [hunk 2]
```diff
@@ -286,3 +290,26 @@ def initialize_dummy_weights(
     for param in model.state_dict().values():
         if torch.is_floating_point(param):
             param.data.uniform_(low, high)
+
+
+def load_tensorized_weights(model: torch.nn.Module,
+                            model_config: ModelConfig):
+    before_mem = get_mem_usage()
+    # Lazy load the tensors from S3 into the model.
+    start = time.time()
```
Comment (dmarx) [inline vllm/model_executor/weight_utils.py]: the timing stuff and the bragging aren't needed here, they're just illustrative in the tensorizer example.

File: vllm/model_executor/weight_utils.py [hunk 3]
```diff
@@ -12,7 +13,10 @@
 import torch
 from transformers import PretrainedConfig
 from tqdm.auto import tqdm
+from tensorizer import TensorDeserializer, stream_io
+from tensorizer.utils import convert_bytes, get_mem_usage
```
Comment (dmarx) [inline vllm/model_executor/weight_utils.py]: don't need to import convert_bytes or get_mem_usage

File: vllm/model_executor/weight_utils.py [hunk 4]
```diff
@@ -286,3 +290,26 @@ def initialize_dummy_weights(
     for param in model.state_dict().values():
         if torch.is_floating_point(param):
             param.data.uniform_(low, high)
+
+
+def load_tensorized_weights(model: torch.nn.Module,
+                            model_config: ModelConfig):
+    before_mem = get_mem_usage()
+    # Lazy load the tensors from S3 into the model.
```
Comment (dmarx) [inline vllm/model_executor/weight_utils.py]: i think highlighting the lazy loading piece is helpful for people who may be unfamiliar with tensorizer. this should accompany the `no_init_or_tensor` invocation in `model_loader.get_model`

**Review by dmarx — CHANGES_REQUESTED** [review]
mostly lgtm. should probably remove the timing and memory benchmarking bits from `load_tensorized_weights`. If we really want to keep that stuff in, let's emit those messages using vllm's logger to facilitate end user control over verbosity

```python
from vllm.logger import init_logger

logger = init_logger(__name__)
```

EDIT: better yet, if we want to keep those messages, we could push up the timing and memory stuff so it will be applied agnostically to how the model was loaded. this way, users can see how utilizing tensorizer changes things.

File: vllm/config.py (hunk 1 above)
Comment (sangstar) [reply vllm/config.py]: I like that idea. I imagine this would increase the overhead of `tensorizer` literacy for `vllm` users, if this were to hypothetically not just be used by our team internally? My aim for this was to make it agnostic of `tensorizer`'s library and only need to have tensorized weights findable, but that requires tensorizing weights in the first place, so some understanding of how `tensorizer` works seems to be necessary

File: vllm/model_executor/weight_utils.py (hunk 2 above)
Comment (sangstar) [reply vllm/model_executor/weight_utils.py]: This is fair. I mainly included this for myself to make sure everything was working and was debating whether it was useful to display for anyone else.

File: vllm/config.py (hunk 1 above)
Comment (dmarx) [reply vllm/config.py]: yeah i'm not in love with my proposal here. I do think it is important to at least expose the encryption functionality. Another path forward here would be to modify `get_model` to optionally take a `TensorizerConfig` that would encapsulate all the tensorizer-specific stuff. It also just feels weird sprinkling all these tensorizer-specific things in the vllm APIs. Maybe it would make sense to have a separate `get_tensorized_model` function that could handle tensorizer-specific arguments?

File: vllm/config.py [hunk 5]
```diff
@@ -415,6 +423,7 @@ def _get_and_verify_dtype(
     # NOTE: getattr(config, "torch_dtype", torch.float32) is not correct
     # because config.torch_dtype can be None.
     config_dtype = getattr(config, "torch_dtype", None)
+    print(f"dtype given from config is {config_dtype}")
```
Comment (rtalaricw) [inline vllm/config.py]: Debug print statement?

File: vllm/engine/arg_utils.py [hunk 6]
```diff
@@ -1,12 +1,112 @@
...
+        )
+
+    @staticmethod
+    def add_cli_args(
+            parser: argparse.ArgumentParser) -> argparse.ArgumentParser:
+        """Tensorizer CLI arguments"""
+        parser.add_argument(
+            "--serialize",
+            action='store_true',
+            help="In the event that serialized weights can't be found at"
```
Comment (rtalaricw) [inline vllm/engine/arg_utils.py]: Seems like all help descriptions are the same.

**Comment by rtalaricw** [comment]
If the model does not exist to be de-serialized, we should serialize the model to the appropriate path (local/S3) and then de-serialize it, right? Is this what is currently happening?

**Comment by rtalaricw** [comment]
> If the model does not exist to be de-serialized, we should serialize the model to the appropriate path (local/S3) and then de-serialize it, right? Is this what is currently happening?

Should we automate that or let the user serialize independently?

File: vllm/config.py [hunk 7]
```diff
@@ -55,6 +58,9 @@ class ModelConfig:
         max_context_len_to_capture: Maximum context len covered by CUDA graphs.
             When a sequence has context length larger than this, we fall back
             to eager mode.
+        serialize: In the event that serialized weights can't be found at
```
Comment (dmarx) [inline vllm/config.py:60]: in retrospect: adding this will make the tensorizer behavior inconsistent with the default huggingface transformers behavior, which will save the weights to the download_dir whether or not this argument is here. additionally: I don't think this belongs in the model config.

File: vllm/engine/arg_utils.py (hunk 6 above)
Comment (sangstar) [reply vllm/engine/arg_utils.py]: Yep, I'm going to get to that

File: vllm/config.py (hunk 5 above)
Comment (sangstar) [reply vllm/config.py]: Yeah I'm not at the cleaning/formatting stage just yet

**Comment by dmarx** [comment]
> If the model does not exist to be de-serialized, we should serialize the model to the appropriate path (local/S3) and then de-serialize it, right? Is this what is currently happening?
Sanger and I had some discussion offline that I should have summarized here for your visibility. Here's the plan we came up with:

* rather than adding tensorizer-specific arguments to the model config, we interpret `download_path` as "expected location where the model should be found" and support this being either a local path or remote object storage (e.g. s3 bucket) if the specified `load_format` is "tensorizer"
* if the expected file is not found at the `download_path`, tensorizer will serialize it for the user. To give users additional control over this behavior, serialization will only happen if a `--serialize` argument is provided, otherwise a FileNotFound error will be thrown with a message instructing the user to use the `--serialize` argument. 
  * In retrospect i'm thinking this is unnecessary and worse inconsistent with vllm's current behavior. if we really want to do this, we should make the `--serialize` argument impact huggingface transformers save behavior to the `download_dir` as well. if we want, we can do this in a separate PR which we could stack on top of this, but I imagine this would probably be rejected since it would likely be a breaking change for a lot of vllm users. if people aren't complaining about accidentally downloading files to misspelled file paths, we shouldn't fix a problem that isn't there.
* tensorizer-specific arguments (e.g. encryption params) to be optionally provided via a to-purpose object that will encapsulate these arguments. preferred behavior will be to provide them via a config file specific to this argument group

**Comment by sangstar** [comment]
Made some changes:

- `--serialize` has been removed to conform with HuggingFace's practice of automatically saving model weights at `download_dir` if no model weights can be found 
- Updated CLI arg descriptions

Currently implemented passing args for serialization and deserialization by passing a `tensorizer_args` param to `EngineArgs` and `ModelConfig`, as I felt this was less intrusive than creating an entirely new `TensorizerConfig` object, and `TensorizerAgent` makes use of `ModelConfig` anyway

**Comment by rtalaricw** [comment]
> > If the model does not exist to be de-serialized, we should serialize the model to the appropriate path (local/S3) and then de-serialize it, right? Is this what is currently happening?
> > Sanger and I had some discussion offline that I should have summarized here for your visibility. Here's the plan we came up with:
> 
> * rather than adding tensorizer-specific arguments to the model config, we interpret `download_path` as "expected location where the model should be found" and support this being either a local path or remote object storage (e.g. s3 bucket) if the specified `load_format` is "tensorizer"
> * if the expected file is not found at the `download_path`, tensorizer will serialize it for the user. To give users additional control over this behavior, serialization will only happen if a `--serialize` argument is provided, otherwise a FileNotFound error will be thrown with a message instructing the user to use the `--serialize` argument.
>   
>   * In retrospect i'm thinking this is unnecessary and worse inconsistent with vllm's current behavior. if we really want to do this, we should make the `--serialize` argument impact huggingface transformers save behavior to the `download_dir` as well. if we want, we can do this in a separate PR which we could stack on top of this, but I imagine this would probably be rejected since it would likely be a breaking change for a lot of vllm users. if people aren't complaining about accidentally downloading files to misspelled file paths, we shouldn't fix a problem that isn't there.
> * tensorizer-specific arguments (e.g. encryption params) to be optionally provided via a to-purpose object that will encapsulate these arguments. preferred behavior will be to provide them via a config file specific to this argument group

So a user would:


`python3 -m vllm.entrypoints.openai.api_server \
--host "0.0.0.0" \
--model "EleutherAI/gpt-j-6b" --load-format "tensorizer" \
--tensorizer-path "s3://tensorized-ssteel/gpt-j-6b-vllm.tensors" --serialize` 

and then call:

`python3 -m vllm.entrypoints.openai.api_server \
--host "0.0.0.0" \
--model "EleutherAI/gpt-j-6b" --load-format "tensorizer" \
--tensorizer-path "s3://tensorized-ssteel/gpt-j-6b-vllm.tensors"`

again?

Why can't we provide a tensorizer_utils file to perform any additional work necessary or even leave serialization out since users would have to go through tensorizer and understand the basics to be able to leverage its functionality here?

**Comment by sangstar** [comment]
> > > If the model does not exist to be de-serialized, we should serialize the model to the appropriate path (local/S3) and then de-serialize it, right? Is this what is currently happening?
> > > Sanger and I had some discussion offline that I should have summarized here for your visibility. Here's the plan we came up with:
> > 
> > 
> > 
> > * rather than adding tensorizer-specific arguments to the model config, we interpret `download_path` as "expected location where the model should be found" and support this being either a local path or remote object storage (e.g. s3 bucket) if the specified `load_format` is "tensorizer"
> > * if the expected file is not found at the `download_path`, tensorizer will serialize it for the user. To give users additional control over this behavior, serialization will only happen if a `--serialize` argument is provided, otherwise a FileNotFound error will be thrown with a message instructing the user to use the `--serialize` argument.
> >   
> >   * In retrospect i'm thinking this is unnecessary and worse inconsistent with vllm's current behavior. if we really want to do this, we should make the `--serialize` argument impact huggingface transformers save behavior to the `download_dir` as well. if we want, we can do this in a separate PR which we could stack on top of this, but I imagine this would probably be rejected since it would likely be a breaking change for a lot of vllm users. if people aren't complaining about accidentally downloading files to misspelled file paths, we shouldn't fix a problem that isn't there.
> > * tensorizer-specific arguments (e.g. encryption params) to be optionally provided via a to-purpose object that will encapsulate these arguments. preferred behavior will be to provide them via a config file specific to this argument group
> 
> So a user would:
> 
> `python3 -m vllm.entrypoints.openai.api_server \ --host "0.0.0.0" \ --model "EleutherAI/gpt-j-6b" --load-format "tensorizer" \ --tensorizer-path "s3://tensorized-ssteel/gpt-j-6b-vllm.tensors" --serialize`
> 
> and then call:
> 
> `python3 -m vllm.entrypoints.openai.api_server \ --host "0.0.0.0" \ --model "EleutherAI/gpt-j-6b" --load-format "tensorizer" \ --tensorizer-path "s3://tensorized-ssteel/gpt-j-6b-vllm.tensors"`
> 
> again?
> 
> Why can't we provide a tensorizer_utils file to perform any additional work necessary or even leave serialization out since users would have to go through tensorizer and understand the basics to be able to leverage its functionality here?

Users no longer have to provide `--serialize`; if no serialized tensors are found at `download_dir`, model weights will be downloaded from HF and serialized to `download_dir` automatically. Currently, this stops the process and providing the same CLI command will then deserialize properly. I should probably instead allow the downloaded model weights from HF then be used to serve the model rather than having the process restart. I'll add that to the list. There's no additional technical overhead for users to serialize tensors automatically if trying to use `tensorizer`

**Comment by dmarx** [comment]
yeah i thought the way this would work was if the file doesn't exist at the `download_path` and the `download_path` is remote, the model would be downloaded to the default HF cache, loaded into RAM from there, serialized to the remote path, (~~deleted from the HF cache?~~ EDIT: yeah nm that's way overcomplicated), and then deserializing from the remote shouldn't be necessary because the model was already loaded locally in order to serialize it in the first place

**Comment by dmarx** [comment]
> Why can't we provide a tensorizer_utils file to perform any additional work necessary or even leave serialization out since users would have to go through tensorizer and understand the basics to be able to leverage its functionality here?

my hope is that we can deploy this feature in a way that will let users leverage tensorizer without knowing basically anything about it. user just provides the object storage location, and the implementation handles accelerating their model loading w/tensorizer whether they've put the model there already or not. if the user wants to leverage more advanced features of tensorizer like encryption, we expose those features through config arguments. But hopefully we can set this up so the majority of users can leverage the benefits of incorporating tensorizer with minimal overhead

**Comment by sangstar** [comment]
At this time, users can leverage `tensorizer` just by specifying `--load-format "tensorizer"`. Everything else is handled by default as long as `download_dir` points to a local filepath or an S3 endpoint. The user only has to know that `tensorizer` works by deserializing serialized model weights at `download_dir`, and if they don't have that beforehand, it'll be done for them.

If they want, they can finetune further by passing additional params in the CLI, but all they technically need to know is that `download_dir` should theoretically point to serialized tensors, and if not, they will be created. 

The understanding that "`tensorizer` uses their own special form of model weights to load models super fast. I just need to provide a path where the special model weights are found to load them for my service, or I can specify a path for it to be saved if I don't have them" is all the technical expertise required to use it

File: vllm/model_executor/model_loader.py [hunk 8]
```diff
@@ -1,16 +1,107 @@
...
+                    f"Tensors not found. Will load via HF and serialize tensors to {self.tensorizer_args.download_dir}"
+                )
+                return False
+            else:
+                raise OSError(err)
+
+    def serialize(self):
+        with torch.device("cuda"):
+            model = self.model_cls(self.model_config.hf_config)
+        self.model_config.load_format = "auto"
```
Comment (rtalaricw) [inline vllm/model_executor/model_loader.py]: Were you able to try `fp16` or anything that is not `auto`? What was the issue here?

**Review by rtalaricw — COMMENTED** [review]
LGTM. I think formatting it using black/isort and submitting as a PR to the vllm main branch would be a good next step to get feedback on what they think is required to integrate it as a feature in the vllm repository for more awareness and evangelism of tensorizer. @dmarx I am curious what you think?

File: vllm/model_executor/model_loader.py (hunk 8 above)
Comment (sangstar) [reply vllm/model_executor/model_loader.py]: `load_format` does not specify the `dtype` of the loaded model. It specifies the way in which model weights should be loaded, whether that's using `auto`, `pt`, `safetensors`, `npcache`, `dummy`, and now `tensorizer`. In the case of serializing model weights if no weights can be found at `download_dir`, this logic is to support loading the model through other means to then serialize it. `auto` is the default load format option, and seemed appropriate here. It attempts to load the model using `safetensors`, and failing that, through `pt`.

**Comment by sangstar** [comment]
Update to the code:

**HF models can now be deserialized.** Deserializing now directly baked into `hf_model_weights_iterator()` to allow for the HF-vLLM model translation performed during `load_weights`. Deserialization args can be parsed as normal, **but support is now removed for automatically serializing weights if weights not located at `download_dir`.**

**Comment by dmarx** [comment]
LPT: instead of merging the upstream changes, if you rebase on top of them instead they won't get folded into your PR (e.g. the PR currently says you've changed 96 files, which obviously isn't correct)

**Comment by dmarx** [comment]
NB: current implementation requires users to provide s3 credentials

TODO: users should be able to download the public pre-tensorized checkpoints into vllm un-credentialled (http). Tensorizer already supports this.

**Comment by lizzzcai** [comment]
Hi, I noticed this repo and would like to check if it is ready for testing? and is there a plan to support it in the official vllm, thanks

**Comment by sangstar** [comment]
> Hi, I noticed this repo and would like to check if it is ready for testing? and is there a plan to support it in the official vllm, thanks

Hi there! Work has migrated to the `sangstar/integrate-tensorizer` branch which you can track. During testing, we noticed that tensors were double-copied in GPU memory, as vLLM allocates an initialized model with zero tensors and loads the yielded weights onto those tensors before copying it to the GPU. `tensorizer`'s `TensorDeserializer` handles the loading of tensors into GPU memory already, so work is currently being done to load the yielded weights for their bespoke models classes while preventing vLLM from intervening and double-copying.

**Comment by lizzzcai** [comment]
> > Hi, I noticed this repo and would like to check if it is ready for testing? and is there a plan to support it in the official vllm, thanks
> 
> Hi there! Work has migrated to the `sangstar/integrate-tensorizer` branch which you can track. During testing, we noticed that tensors were double-copied in GPU memory, as vLLM allocates an initialized model with zero tensors and loads the yielded weights onto those tensors before copying it to the GPU. `tensorizer`'s `TensorDeserializer` handles the loading of tensors into GPU memory already, so work is currently being done to load the yielded weights for their bespoke models classes while preventing vLLM from intervening and double-copying.

Thanks @sangstar for your reply. I did a quick test today, and found out that some models are working like `bigscience/bloom-560m`, but when I tried out with `mistralai/Mistral-7B-Instruct-v0.1`, I got this error: `RuntimeError: The size of tensor a (0) must match the size of tensor b (4096) at non-singleton dimension 2` Testing with Tensorizer alone is working fine with Mistral model.

**Comment by sangstar** [comment]
> > > Hi, I noticed this repo and would like to check if it is ready for testing? and is there a plan to support it in the official vllm, thanks
> > 
> > 
> > Hi there! Work has migrated to the `sangstar/integrate-tensorizer` branch which you can track. During testing, we noticed that tensors were double-copied in GPU memory, as vLLM allocates an initialized model with zero tensors and loads the yielded weights onto those tensors before copying it to the GPU. `tensorizer`'s `TensorDeserializer` handles the loading of tensors into GPU memory already, so work is currently being done to load the yielded weights for their bespoke models classes while preventing vLLM from intervening and double-copying.
> 
> Thanks @sangstar for your reply. I did a quick test today, and found out that some models are working like `bigscience/bloom-560m`, but when I tried out with `mistralai/Mistral-7B-Instruct-v0.1`, I got this error: `RuntimeError: The size of tensor a (0) must match the size of tensor b (4096) at non-singleton dimension 2` Testing with Tensorizer alone is working fine with Mistral model.

Is this with this branch or with `sangstar/integrate-tensorizer`?

**Comment by lizzzcai** [comment]
Hi @sangstar , yes, I am testing on your branch.

**Comment by sangstar** [comment]
> Hi @sangstar , yes, I am testing on your branch.

`sangstar/integrate-tensorizer`'s recent commit when you were testing was a WIP commit and isn't meant to be at a stable state. I will be moving the design philosophy likely to be focused primarily on deserializing vLLM-formatted models, and will likely create a PR on `sangstar/integrate-tensorizer` when it's stable. 

The most recent commit on that branch is at a more stable state, although it's still in development and I haven't tested it thoroughly.
